apps/dashboard/pkg/migration/conversion/datasource.go

```diff
@@ -0,0 +1,133 @@
+package conversion
+
+import (
+	"context"
+
+	dashv2alpha1 "github.com/grafana/grafana/apps/dashboard/pkg/apis/dashboard/v2alpha1"
+	"github.com/grafana/grafana/apps/dashboard/pkg/migration/schemaversion"
+)
+
+// DataSourceProvider interface for getting datasource information
+type DataSourceProvider interface {
+	GetDataSourceByUID(uid string) *DataSourceInfo
+	GetDefaultDataSource() DataSourceInfo
+}
+
+// DataSourceInfo contains information about a datasource
+type DataSourceInfo struct {
+	UID  string
+	Type string
+}
+
+// DataSourceProviderAdapter adapts the schemaversion.DataSourceInfoProvider to the conversion DataSourceProvider
+type DataSourceProviderAdapter struct {
+	provider  schemaversion.DataSourceInfoProvider
+	dsMap     map[string]schemaversion.DataSourceInfo
+	defaultDS *schemaversion.DataSourceInfo
+}
+
+// NewDataSourceProviderAdapter creates a new adapter
+func NewDataSourceProviderAdapter(provider schemaversion.DataSourceInfoProvider) *DataSourceProviderAdapter {
+	dsMap := make(map[string]schemaversion.DataSourceInfo)
+	var defaultDS *schemaversion.DataSourceInfo
+
+	// Pre-load all datasources into a map for quick lookup
+	datasources := provider.GetDataSourceInfo(context.Background())
+	for _, ds := range datasources {
+		dsMap[ds.UID] = ds
+		if ds.Default {
+			dsRef := ds
+			defaultDS = &dsRef
+		}
+	}
+
+	// If no default was found, fallback to Grafana
+	if defaultDS == nil {
+		defaultDS = &schemaversion.DataSourceInfo{
+			UID:  "-- Grafana --",
+			Type: "grafana",
+		}
+	}
+
+	return &DataSourceProviderAdapter{
+		provider:  provider,
+		dsMap:     dsMap,
+		defaultDS: defaultDS,
+	}
+}
+
+// GetDataSourceByUID returns datasource info for a given UID
+func (a *DataSourceProviderAdapter) GetDataSourceByUID(uid string) *DataSourceInfo {
+	if ds, ok := a.dsMap[uid]; ok {
+		return &DataSourceInfo{
+			UID:  ds.UID,
+			Type: ds.Type,
+		}
+	}
+	return nil
+}
+
+// GetDefaultDataSource returns the default datasource info
+func (a *DataSourceProviderAdapter) GetDefaultDataSource() DataSourceInfo {
+	return DataSourceInfo{
+		UID:  a.defaultDS.UID,
+		Type: a.defaultDS.Type,
+	}
+}
+
+var globalDataSourceProvider DataSourceProvider
```

Comment:
we'll need to find another way to do this - this wont work for the standalone apiserver

Reply:
What about passing it through the context? Opening a PR on top of this one to show you :)